src/com/sun/jna/Cleaner.java

```diff
@@ -0,0 +1,152 @@
+                        Reference<? extends Object> ref = referenceQueue.remove();
+                        if(ref instanceof CleanerRef) {
+                            ((CleanerRef) ref).clean();
+                        }
+                    } catch (InterruptedException ex) {
+                        Logger.getLogger(Cleaner.class.getName()).log(Level.SEVERE, null, ex);
+                        break;
+                    } catch (Exception ex) {
+                        Logger.getLogger(Cleaner.class.getName()).log(Level.SEVERE, null, ex);
+                    }
+                }
+            }
+        };
+        cleanerThread.setName("JNA Cleaner");
+        cleanerThread.setDaemon(true);
+        cleanerThread.start();
+    }
+
+    public synchronized CleanerRef register(Object obj, Runnable cleanupTask) {
+        // The important side effect is the PhantomReference, that is yielded
+        // after the referent is GCed
+        return add(new CleanerRef(this, obj, referenceQueue, cleanupTask));
+    }
+
+    private synchronized CleanerRef add(CleanerRef ref) {
+        if(firstCleanable == null) {
+            firstCleanable = ref;
+        } else {
+            ref.setNext(firstCleanable);
+            firstCleanable.setPrevious(ref);
+            firstCleanable = ref;
+        }
+        return ref;
+    }
+
+    private synchronized boolean remove(CleanerRef ref) {
+        boolean inChain = false;
+        if(ref == firstCleanable) {
+            firstCleanable = ref.getNext();
+            inChain = true;
+        }
+        if(ref.getPrevious() != null) {
+            ref.getPrevious().setNext(ref.getNext());
+        }
+        if(ref.getNext() != null) {
+            ref.getNext().setPrevious(ref.getPrevious());
+        }
+        if(ref.getPrevious() != null || ref.getNext() != null) {
+            inChain = true;
+        }
+        ref.setNext(null);
+        ref.setPrevious(null);
+        return inChain;
+    }
+
+    private static class CleanerRef extends PhantomReference<Object> implements Cleanable {
+        private final Cleaner cleaner;
+        private final Runnable cleanupTask;
+        private CleanerRef previous;
+        private CleanerRef next;
+
+        public CleanerRef(Cleaner cleaner, Object referent, ReferenceQueue<? super Object> q, Runnable cleanupTask) {
+            super(referent, q);
+            this.cleaner = cleaner;
+            this.cleanupTask = cleanupTask;
+        }
+
+        public void clean() {
+            if(cleaner.remove(this)) {
+                cleanupTask.run();
+            }
+        }
+
+        CleanerRef getPrevious() {
+            return previous;
+        }
+
+        void setPrevious(CleanerRef previous) {
+            this.previous = previous;
+        }
+
+        CleanerRef getNext() {
+            return next;
+        }
+
+        void setNext(CleanerRef next) {
+            this.next = next;
+        }
+    }
+
+    public static interface Cleanable {
```

Comment:
Since this uses the same name as a JDK9+ class, recommend a javadoc indicating that it's intended to serve the same purpose.  (I assume you based your code on the JDK9 Cleaner?)

Reply:
Yes this is modelled after the JDK9+ Cleaner. The cleaner is intended to be used by JNA internal code. I made this cleaner in the update by moving the cleaner into the "internal" package and adjust the documentation slightly.